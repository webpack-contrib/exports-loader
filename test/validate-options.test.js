import { getCompiler, compile } from './helpers';

describe('validate options', () => {
  const tests = {
    type: {
      success: ['commonjs', 'module'],
      failure: [true, false, 'foo'],
    },
    exports: {
      success: [
        'Foo',
        'default Foo',
        'named Foo',
        'named Foo FooA',
        ['named Foo', 'named Bar'],
        ['named Foo FooA', 'named Bar BarA'],
        ['default Foo', 'named Bar BarA'],
        'single Foo',
        'multiple Foo',
        'multiple Foo FooA',
        ['multiple Foo', 'multiple Bar'],
        ['multiple Foo FooA', 'multiple Bar BarA'],
      ],
      failure: [true, () => {}, [1, 2, 3]],
    },
  };

  function stringifyValue(value) {
    if (
      Array.isArray(value) ||
      (value && typeof value === 'object' && value.constructor === Object)
    ) {
      return JSON.stringify(value);
    }

    return value;
  }

  async function createTestCase(key, value, type) {
    it(`should ${
      type === 'success' ? 'successfully validate' : 'throw an error on'
    } the "${key}" option with "${stringifyValue(value)}" value`, async () => {
      // eslint-disable-next-line no-shadow
      const getOptions = (key, value) => {
        if (key === 'type') {
          return { [key]: value, exports: 'Foo' };
        }

        if (
          key === 'exports' &&
          (typeof value === 'string' ||
            (Array.isArray(value) &&
              value.filter((item) => typeof item !== 'string').length === 0))
        ) {
          // eslint-disable-next-line no-shadow
          const type = Array.isArray(value)
            ? value.filter(
                (item) => item.includes('single') || item.includes('multiple')
              ).length > 0
              ? 'commonjs'
              : 'module'
            : value.includes('single') || value.includes('multiple')
            ? 'commonjs'
            : 'module';

          return { type, [key]: value };
        }

        return { [key]: value };
      };

      const compiler = getCompiler('simple.js', getOptions(key, value));

      let stats;

      try {
        stats = await compile(compiler);
      } finally {
        if (type === 'success') {
          expect(stats.hasErrors()).toBe(false);
        } else if (type === 'failure') {
          const {
            compilation: { errors },
          } = stats;

          expect(errors).toHaveLength(1);
          expect(() => {
            throw new Error(errors[0].error.message);
          }).toThrowErrorMatchingSnapshot();
        }
      }
    });
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values)) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
