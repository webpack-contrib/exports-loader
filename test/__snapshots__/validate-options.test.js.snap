// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate options should throw an error on the "exports" option with "() => {}" value 1`] = `
"Invalid options object. Exports Loader has been initialized using an options object that does not match the API schema.
 - options.exports should be one of these:
   non-empty string | [non-empty string, ...] (should not have fewer than 1 item)
   Details:
    * options.exports should be a non-empty string.
    * options.exports should be an array:
      [non-empty string, ...] (should not have fewer than 1 item)"
`;

exports[`validate options should throw an error on the "exports" option with "[1,2,3]" value 1`] = `
"Invalid options object. Exports Loader has been initialized using an options object that does not match the API schema.
 - options.exports should be one of these:
   non-empty string | [non-empty string, ...] (should not have fewer than 1 item)
   Details:
    * options.exports[2] should be a non-empty string."
`;

exports[`validate options should throw an error on the "exports" option with "true" value 1`] = `
"Invalid options object. Exports Loader has been initialized using an options object that does not match the API schema.
 - options.exports should be one of these:
   non-empty string | [non-empty string, ...] (should not have fewer than 1 item)
   Details:
    * options.exports should be a non-empty string.
    * options.exports should be an array:
      [non-empty string, ...] (should not have fewer than 1 item)"
`;

exports[`validate options should throw an error on the "type" option with "false" value 1`] = `
"Invalid options object. Exports Loader has been initialized using an options object that does not match the API schema.
 - options.type should be one of these:
   \\"commonjs\\" | \\"module\\""
`;

exports[`validate options should throw an error on the "type" option with "foo" value 1`] = `
"Invalid options object. Exports Loader has been initialized using an options object that does not match the API schema.
 - options.type should be one of these:
   \\"commonjs\\" | \\"module\\""
`;

exports[`validate options should throw an error on the "type" option with "true" value 1`] = `
"Invalid options object. Exports Loader has been initialized using an options object that does not match the API schema.
 - options.type should be one of these:
   \\"commonjs\\" | \\"module\\""
`;
